libcloud/dns/drivers/cloudflare.py

```diff
@@ -141,290 +159,252 @@ class CloudFlareDNSDriver(DNSDriver):
+                                                  record.id)
 
-        for item in data:
-            record = self._to_record(zone=zone, item=item)
-            records.append(record)
+        self.connection.set_context({'record_id': record.id})
+        response = self.connection.request(url, method='DELETE')
 
-        return records
+        item = response.object.get('result', {}).get('id')
+        is_deleted = item == record.id
 
-    def _to_record(self, zone, item):
-        name = self._get_record_name(item=item)
-        type = item['type']
-        data = item['content']
-
-        if item.get('ttl', None):
-            ttl = int(item['ttl'])
-        else:
-            ttl = None
-
-        extra = {}
-        extra['ttl'] = ttl
-        extra['props'] = item.get('props', {})
-        for attribute in RECORD_EXTRA_ATTRIBUTES:
-            value = item.get(attribute, None)
-            extra[attribute] = value
-
-        record = Record(id=str(item['rec_id']), name=name, type=type,
-                        data=data, zone=zone, driver=self, ttl=ttl,
-                        extra=extra)
-        return record
+        return is_deleted
+
+    def ex_get_user_account_memberships(self):
+        def _ex_get_user_account_memberships(params):
+            url = '{}/memberships'.format(API_BASE)
 
-    def _get_record_name(self, item):
-        name = item['name'].replace('.' + item['zone_name'], '') or None
-        if name:
-            name = name.replace(item['zone_name'], '') or None
-        return name
+            response = self.connection.request(url, params=params)
+            return response.object['result']
+
+        return self._paginate(_ex_get_user_account_memberships)
+
+    def _to_zone(self, item):
+        return Zone(
+            id=item['id'],
+            domain=item['name'],
+            type=CLOUDFLARE_TO_LIBCLOUD_ZONE_TYPE[item['type']],
+            ttl=None,
+            driver=self,
+            extra={key: item.get(key) for key in ZONE_EXTRA_ATTRIBUTES},
+        )
+
+    def _to_record(self, zone, item):
+        name = item['name']
+        name = name.replace('.' + item['zone_name'], '')
+        name = name.replace(item['zone_name'], '')
+        name = name or None
+
+        ttl = item.get('ttl')
+        if ttl is not None:
+            ttl = int(ttl)
+
+        return Record(
+            id=item['id'],
+            name=name,
+            type=item['type'],
+            data=item['content'],
+            zone=zone,
+            driver=self,
+            ttl=ttl,
+            extra={key: item.get(key) for key in RECORD_EXTRA_ATTRIBUTES},
+        )
+
+    def _paginate(self, get_page):
+        for page in itertools.count(start=1):
+            params = {'page': page, 'per_page': self.PAGE_SIZE}
+
+            items = get_page(params)
+
+            for item in items:
+                yield item
+
+            if len(items) < self.PAGE_SIZE:
```

Comment:
Does CloudFlare API return some kind of indication in the response headers which could let us know if there are more results / next page?

If so, this could help us avoid unnecessary HTTP request where we are on the last page, but the last page contains ``PAGE_SIZE`` number of items. That is assuming that assuming that the API doesn't return some kind of error if you request a page which doesn't exist.

Reply:
EDIT:

Yeah, it looks like it should be contain in the optional ``result_info`` attribute in the response body - https://api.cloudflare.com/

Reply:
Good catch. Used the `result_info` object if it's present to avoid the extra HTTP request in 828b471.